### ISSUE TITLE
Thoughts

### ISSUE THREAD

**Issue by OliverJAsh** [jackfranklin/jspm-dev-builder#1]
Couple of thoughts:
1. `inLoc` would be better named as `expression`. jspm defines a bundle expression as something like `main` or `main - jquery` (it can contain arithmetic). https://github.com/jspm/jspm-cli/blob/84fd8cff30ae538970b90c143ec8a7d99f6bf725/lib/bundle.js#L75
2. We should add the ability to pass bundle options (minify, no mangle, source map, source map contents…)
3. We should add the ability to choose between bundle and bundle SFX (`builder.buildStatic`)
4. I'm not sure about the implementation of `removeFromTrace` (https://github.com/jackfranklin/jspm-dev-builder/blob/df28dd35e43968b75cbeed425dcc9c362b1756d3/index.js#L18-L24). If the filename passed in is absolute (as is often for watch tasks), it won't match. Perhaps this function should expect a `moduleId` which it can match exactly in `sharedCache.trace`. Then, it is up to the consumer to map file paths to SystemJS module ID in the watch event listener. Alternatively it could receive the absolute path, make it relative to the jspm base URL, and then run it through `System.normalize` to get the module ID.

For now I have just copied the code into my project for direct use. What do you think?

``` js
/* eslint-env node */
import gulp from 'gulp';
import watch from 'gulp-watch';
import jspm from 'jspm';
import path from 'path';

let builderCache = {};

const removeFromTrace = (moduleId) => delete builderCache.trace[moduleId];

const moduleExpression = 'main';
const writeFileName = `${__dirname}/public/js/main-bundle.js`;

const build = (changedModuleId) => {
    const builder = new jspm.Builder();

    if (changedModuleId) {
        removeFromTrace(changedModuleId);
    }

    builder.setCache(builderCache);

    const buildStart = Date.now();
    return builder.buildStatic(moduleExpression, writeFileName)
        .then(() => {
            builderCache = builder.getCache();

            const buildEnd = Date.now();
            console.log(`bundle time: ${buildEnd - buildStart}`);
        });
};

gulp.task('build', () => build());

gulp.task('watch', ['build'], () => {
    const baseURL = new jspm.Builder().loader.baseURL.replace('file://', '');
    return watch(`${baseURL}/**/*.js`, (vinyl) => {
        // TODO: Map to module ID. How?
        const moduleId = path.relative(baseURL, vinyl.path);
        console.log(`changed: ${moduleId}`);
        build(moduleId);
    });
});
```

**Comment by jackfranklin** [comment]
1. Yep!
2. Yep!
3. Yep!

:+1: 
1. Yeah, the implementation currently is very much built precisely to match what worked for us at GC, and definitely can be made much more robust. We should definitely make use of anything built into SystemJS / jspm that makes this better.

The reason the implementation is how it is at the moment is because a bunch of things in the cache ended up looking like:

```
my/file.html!plugin/text@0.0.2...
```

So we need to account for that as well as the keys being just file paths.

**Comment by stoiet** [comment]
Nice! I agree with all points. I think Systemjs/builder should provide better documentation of the API. I am not sure that i understand all the code of systemjs/builder, but you showed a great example. Thanks!

**Comment by jackfranklin** [comment]
Pinging in @guybedford here too to see what he thinks to the above :+1: 

@stoiet glad this was useful - if there's specific areas of docs you think are missing please let me know, happy to work on some now I'm getting a little more familiar with jspm.

**Comment by jackfranklin** [comment]
@OliverJAsh have opened #3 to track your last point.

All the others have been fixed on master and are pending a 0.2 release.
